[B+C] Implement method to get arrow itemstack and cancel item consume in EntityShootBowEvent. Adds BUKKIT-4686

me4502 opened the pull request:
<b>The Issue:</b>

Currently it is not possible to get the projectiles itemstack or stop it being consumed in EntityShootBowEvent.

<b>Justification for this PR:</b>

This PR adds the above mentioned capabilities to the event, through 3 new methods. getProjectileStack(), setStackConsumeCancelled(boolean), and isStackConsumeCancelled(). This allows these 3 things to be done, which prior to now were just not plausible through the bukkit API.

The use case this was designed for, is CraftBook Elemental Arrows. They were conceptualized many months ago but never saw the light of day due to bukkit limitations. I needed to get lore and names off arrows, and somehow put them onto the appropriate arrow entities. This would allow me to get the stack, so that I can properly do this and finish the feature that has been sitting collecting dust for 6 months.

<b>PR Breakdown:</b>

This PR grabs a copy of the stack of the arrow before it is removed, and creates a clone of it. This clone is then given the count of 1 so that it is only a single arrow. This is then sent to the event.

The consume cancellation uses the 'flag' field, which currently is used by vanilla when the player is in creative or has the infinity enchantment. The PR is also capable or forcing a arrow to be consumed even in creative/infinity enchantment due to this.

<b>Testing Results and Materials:</b>

I created a small plugin that tests the features outlined in this PR. It is available here, http://pastebin.com/ghJ4LUNh

<b>Relevant PR(s):</b>

B-917 - https://github.com/Bukkit/Bukkit/pull/917
CB-1228 - https://github.com/Bukkit/CraftBukkit/pull/1228

<b>JIRA Ticket:</b>

BUKKIT-4686 - https://bukkit.atlassian.net/browse/BUKKIT-4686

me4502 committed:
Add method to get arrow itemstack, and added methods to cancel item
consume. Adds BUKKIT-4686

gmcferrin commented:
This is currently a breaking change. Deprecate the old event and make a new one, rather than simply removing the old event entirely.

turt2live commented on src/main/java/org/bukkit/event/entity/EntityShootBowEvent.java:
```diff
@@ -39,6 +42,15 @@ public ItemStack getBow() {
     }
 
     /**
+     * Gets the projectile ItemStack fired by the bow; is null if the shooter is a skeleton
```
Can this be adapted to allow for skeletons to be handled?

turt2live commented on src/main/java/org/bukkit/event/entity/EntityShootBowEvent.java:
```diff
@@ -13,13 +13,16 @@
+    private boolean consumeCancelled;
 
-    public EntityShootBowEvent(final LivingEntity shooter, final ItemStack bow, final Projectile projectile, final float force) {
+    public EntityShootBowEvent(final LivingEntity shooter, final ItemStack bow, final ItemStack projectileStack, final Projectile projectile, final float force) {
```
Please deprecate the old constructor and add a new one.

Currently this is a breaking change (as @gmcferrin mentioned)

turt2live commented on src/main/java/org/bukkit/event/entity/EntityShootBowEvent.java:
```diff
@@ -65,6 +77,24 @@ public float getForce() {
+     * 
+     * @return if the stack is being consumed
+     */
+    public boolean isStackConsumeCancelled() {
```
In my opinion this should be renamed.

willConsumeStack() for example. or isConsumingStack()

(just my thoughts though)

turt2live commented on src/main/java/org/bukkit/event/entity/EntityShootBowEvent.java:
```diff
@@ -65,6 +77,24 @@ public float getForce() {
+     * 
+     * @param cancel true if you wish to cancel projectile consumption
+     */
+    public void setStackConsumeCancelled(boolean cancel) {
```
see above.

setConsumingStack() for example

turt2live commented:
Please also describe why this is better than using something like this:

``` java
@EventHandler(priority=EventPriority.MONITOR, ignoreCancelled=true)
public void onShoot(EntityShootBowEvent event){
    if(event.getEntity() instanceof Player){
        Player player = (Player) event.getEntity();
        player.getInventory().addItem(new ItemStack(Material.ARROW));
    }
}

```

(edit)

Ignoring the fact that there are no permission checks, etc. And the fact that I directly typed it here... so it may not compile.

me4502 committed:
Implemented suggestions.

turt2live commented on src/main/java/org/bukkit/event/entity/EntityShootBowEvent.java:
```diff
@@ -65,6 +86,24 @@ public float getForce() {
+    }
+
+    /**
+     * Set whether the projectile ItemStack will be consumed in this event; this has no effect to a skeleton
```
Should be "Sets"

me4502 commented:
This is better than that, because this is a simple cancellation. It also is using the vanilla behaviour of removing the arrow, so you can therefore override infinity enchants or creative simply.

me4502 committed:
Add methods to control consumption of arrows in EntityShootBowEvent.
Adds BUKKIT-4686

Getting the ItemStack of an arrow that is being shot from a bow is
currently not possible, this PR allows this to be possible by adding a
method to EntityShootBotEvent that returns the stack in question.

This PR also adds methods to cancel the consumption of the stack. These
methods can override vanilla behaviours that would normally cancel the
consumption too. For example, infinity enchantments or creative mode.

turt2live commented:
Unfortunately, "because it is simpler" will likely not be enough to get it pulled. 

A change like this will (likely) need a lot of defense as to why this should be pulled. This is simply because the action can already be implemented as I outlined above. 

(and I had a dumb moment earlier: I forgot skeletons do not have a limited amount of arrows...)

me4502 commented:
This also allows the developer to access the stack in question, which I can easily defend; as it is currently not possible.

turt2live commented:
Also please update the commit message (this is CC'd to the CraftBukkit side) to have a more descriptive commit message as per the CONTRIBUTING.md. Something like "Add <stuff>. Adds BUKKIT-#### \n Currently <current situation> is not possible. This commit adds the functionality needed to <stuff> by <what you did>."

me4502 commented:
Done.

turt2live commented:
Commit still isn't perfect :(

The first line is missing the brief description before the "Adds BUKKIT-####". 

"Add methods to control consumption of arrows in EntityShootBowEvent. Adds BUKKIT-4686." then the 2 paragraphs you have there should be sufficient. 

The CraftBukkit side would be something like "Implement methods to control consumption of arrows in EntityShootBowEvent. Adds BUKKIT-4686." then the 2 paragraphs

me4502 commented:
Okay, done.

gmcferrin commented:
Why don't you have the consumeCancelled anywhere in the constructor?

me4502 commented:
I thought it'd just be clutter.

turt2live commented:
Here's the feedback I got from another Bukkit member (this applies to both Bukkit and CraftBukkit):
- The implementation of the item stack used (getArrowStack) is not safe. The inventory could change mid-processing and the stack does not represent the stack to begin with (it's a copy with a forced size of 1).
- This does not appear to safely allow for the aforementioned mid-processing inventory changes.
- This PR is missing full testing materials to represent the possible changes (eg: Changing the inventory). What happens if the player's inventory is cleared while the event is being processed? What happens if arrows are added to the slot before the "current" stack? Does the PR actually work?

My suggestions:
- Remove the getArrowStack(). This is clutter that is not needed.
- Verify the integrity of the server if mid-processing inventory changes occur (and prove the safety of the server)
- Provide a more thorough testing plugin.

me4502 commented:
With their response, the arrow stack returned is not supposed to be a modifiable stack they can change and will effect the inventory; it is supposed to just state what is actually being consumed by the event. The testing material properly tests my given use case, the testing they request are not part of the use case.

With getArrowStack() - That is actually the reason I made this PR. It would be incredibly useful and is not clutter as such, it allows people to find what stack was removed.

With verifying - Not necessary with the use case.

More testing - What else needs to be tested? It outputs the stack upon every bow use.

turt2live commented:
Then the method needs to be adapted to allow for modification of the stack and to have the real stack amount returned. The complaint in question was (paraphrased) "The item stack is a mirror and cannot be edited. What's the point of amount=1 anyways?"

The testing is applicable to the entire PR... If someone shoots a bow and their inventory is cleared while the event is processing, what happens? Does the arrow fire? Does the server spit out an error? Does the server continue? What SHOULD be the expected behavior and what happens?

As for testing, see the above paragraph

turt2live commented:
If you feel this can be resolved with the noted comments, please open a new PR with the changes.
